### PR TITLE
Add optional activation quantization

### PR DESCRIPTION
**Pull request by dblalock**
Adds activation quantization to:

* `MemoryOptimizedMLP` in `megablocks/layers/mlp.py`
* `PaddedScatterOp` in `megablocks/ops/padded_scatter.py`

To support this, we also modify:

* `megablocks/layers/arguments.py` for quantization config
* `megablocks/layers/moe.py` to pass through quantization config
* `megablocks/layers/dmoe.py` to pass through quantization config
* `megablocks/layers/dmoe_test.py` to make sure forward and backward run with quantization. Also, _we now test with `memory_optimized_mlp=True`_.
* `megablocks/ops/padded_scatter_test.py` to make sure forward and backward run with quantization.
* `requirements.txt` and `setup.py` to add the quantization library as a dependency.

[Wandb report](https://wandb.ai/mosaic-ml/davis-actq/reports/Activation-Quantization-Summary--Vmlldzo1Mzg4ODAz) with some results. tl;dr it seems to work fine as long as the GELU inputs are saved as 8b, since these influence the dgrads the most. The FFN inputs and padded_scatter inputs can be 4b. *But* we have a ton of other activations that aren't getting quantized, so we're nowhere near 2x memory savings yet.

### CODE REVIEW

File: requirements.txt
Comment (tgale96) [inline requirements.txt:1]: Could you also update setup.py to install turbo?

File: megablocks/ops/padded_scatter.py [hunk 1]
```diff
@@ -1,17 +1,23 @@
...
     @staticmethod
     @custom_fwd
-    def forward(ctx, x, indices, bin_ids, weights, bins, padded_bins, top_k):
+    def forward(ctx, x, indices, bin_ids, weights, bins, padded_bins, top_k, num_bits):
         saved_x = None if weights is None else x
-        ctx.save_for_backward(saved_x, indices, bin_ids, weights, bins, padded_bins)
+        if num_bits == -1:
+            save_inputs = (saved_x,)
+        else:
+            save_inputs = turbo.quantize_signed(saved_x, num_bits=num_bits)
```
Comment (tgale96) [inline megablocks/ops/padded_scatter.py:16]: I think you need to handle the case where `saved_x` is None?

File: megablocks/layers/dmoe.py [hunk 2]
```diff
@@ -20,6 +20,7 @@ def __init__(self, args : Arguments):
         self.hidden_size = args.hidden_size
         self.ffn_hidden_size = mpu.features_per_rank(args)
         self.blocking = 128
+        self.actq_num_bits = args.quantize_activations_num_bits
```
Comment (tgale96) [inline megablocks/layers/dmoe.py:23]: nit: this is only reused once, so I'd just pass it from self.args instead of creating another member.

File: megablocks/layers/mlp.py [hunk 3]
```diff
@@ -153,17 +154,43 @@ class MemoryOptimizedMLP(torch.autograd.Function):
 
     @staticmethod
     @torch.cuda.amp.custom_fwd
-    def forward(ctx, x, w1, w2, topo):
+    def forward(ctx, x, w1, w2, topo, num_bits):
         # x: [m, k], w1: [n, k], w2: [n, k]
         if (not x.is_contiguous() or not w1.is_contiguous() or
             not w2.is_contiguous()):
             raise ValueError("Expected contiguous 'x', 'w1' and 'w2'.")
+        if num_bits not in (-1, 4, 8):
```
Comment (tgale96) [inline megablocks/layers/mlp.py:162]: Can we do this error checking more centrally? You could put it in a helper and then validate in the MoE/dMoE layers constructors?

File: megablocks/layers/mlp.py [hunk 4]
```diff
@@ -153,17 +154,43 @@ class MemoryOptimizedMLP(torch.autograd.Function):
...
+        if num_bits == -1:
+            gelu_out = gelu.gelu(sdd_out)
+            input_save_args = (x, sdd_out.data)
+        else:
+            # safe version: allocate new mem for the gelu output
+            hidden_q, hidden_scales, gelu_out_data = turbo.quantize_signed(
+                sdd_out.data, num_bits=num_bits, op=turbo.ElemwiseOps.GELU_FORWARD)
+            gelu_out = stk.Matrix(sdd_out.shape, gelu_out_data, *topo_tensors)
+
+            # # bold version: lie about __restrict__ and do the GELU in-place
```
Comment (tgale96) [inline megablocks/layers/mlp.py:186]: Fwiw, I lie about __restrict__ all the time. You just have to make sure your kernel maintains correctness if the pointers are equal, since the compiler won't help you. If you do this in turbo, you could expose an 'in-place' API as well so that the caller doesn't have to worry about the 'boldness'.

File: megablocks/layers/mlp.py [hunk 5]
```diff
@@ -213,7 +248,18 @@ def backward(ctx, ddsd_out):
         # Compute dsdd_out.
         #
         # NOTE: This reuses the dgelu_out allocation.
-        dsdd_out = gelu.gelu_backward_(dgelu_out, sdd_out)
+        if num_bits == -1:
+            dsdd_out = gelu.gelu_backward_(dgelu_out, sdd_out)
+        else:
+            # confusingly, x_out is interpreted as the gradient to overwrite
```
Comment (tgale96) [inline megablocks/layers/mlp.py:254]: lol

**Comment by tgale96** [comment]
Keep me posted on when/if you want me to take a look!

**Comment by dblalock** [comment]
Taking a look now would be good

File: megablocks/layers/dmoe_test.py [hunk 6]
```diff
@@ -137,7 +161,8 @@ def testdMoE_ForwardVersusMoE(
         out, _ = dmoe_mlp(x)
         self.assertSequenceEqual(out.shape, x.shape)
         self.assertSequenceEqual(expected_out.shape, x.shape)
-        self.assertTrue(testing.allclose(out, expected_out))
+        if num_remat_bits == -1:
```
Comment (tgale96) [inline megablocks/layers/dmoe_test.py:164]: I'm assuming this was failing when using quantization? Can we loosen the tolerance for quantized tests, rather than disable completely?

File: megablocks/layers/mlp.py [hunk 7]
```diff
@@ -173,47 +196,82 @@ def forward(ctx, x, w1, w2, topo):
...
     def backward(ctx, ddsd_out):
-        x, w1, w2 = ctx.saved_tensors[:3]
-        sdd_out = stk.Matrix(ctx.shape, *ctx.saved_tensors[3:])
-
         if (not ctx.needs_input_grad[0] or
             not ctx.needs_input_grad[1] or
             not ctx.needs_input_grad[2]):
             raise ValueError("Expected all MLP inputs to need grad.")
 
+        # unpack saved tensors; ugly because quantizing changes tensor count
```
Comment (tgale96) [inline megablocks/layers/mlp.py:215]: Can you make this less ugly by putting the tensors for which there are potentially a variable number at the end of the saved_tensors? The particular part of the logic below that's hard to follow is the modification of saved_tensors.

File: megablocks/ops/padded_scatter.py [hunk 8]
```diff
@@ -20,18 +32,36 @@ def forward(ctx, x, indices, bin_ids, weights, bins, padded_bins, top_k):
...
+                padded_bins,
+                ctx.top_k)
 
         wgrad = None
-        if ctx.needs_input_grad[3]:
+        if ctx.needs_input_grad[3]:  # need wgrad
+            if ctx.num_bits == -1:
+                x = ctx.saved_tensors[0]
+            else:
+                if len(ctx.saved_tensors) != 7:
```
Comment (tgale96) [inline megablocks/ops/padded_scatter.py:52]: What is going on here?

**Comment by tgale96** [comment]
Overall looks good! Have you trained anything with it yet?

File: megablocks/ops/padded_scatter.py (hunk 8 above)
Comment (dblalock) [reply megablocks/ops/padded_scatter.py:52]: I got a weird error indicating that only one tensor got saved in the forward pass when the backward pass was expecting both the quantized input and the scales. I _think_ I fixed this by just checking for `ctx.needs_input_grad[3]` in `saved_x = x if ctx.needs_input_grad[3] else None` instead of for this and `weight is not None`. But I could never repro the error so I added this to at least throw something more informative if it ever happens again. A little ugly and not certain it's needed, so happy to take this block out if you don't like it.

File: megablocks/layers/mlp.py (hunk 7 above)
Comment (dblalock) [reply megablocks/layers/mlp.py:215]: think this is prettier know but I'll leave it up to you as the reader whether this is "resolved"

**Comment by dblalock** [comment]
As far as results: Yes, see the [Wandb report](https://wandb.ai/mosaic-ml/davis-actq/reports/Activation-Quantization-Summary--Vmlldzo1Mzg4ODAz) (let me know if it's not as public as I think it is).

**Comment by tgale96** [comment]
I can't see it - maybe you need to share with my email?

File: megablocks/ops/padded_scatter.py (hunk 8 above)
Comment (tgale96) [reply megablocks/ops/padded_scatter.py:52]: Interesting. I'd say lets remove this if you've never repro'd it and we can add it back if this issue returns.

File: megablocks/layers/dmoe_test.py [hunk 9]
```diff
@@ -146,8 +146,11 @@ def testdMoE_ForwardVersusBaseline(self, bs, sl, hs):
     def testdMoE_ForwardVersusMoE(
             self, bs, sl, hs, num_experts, top_k,
             num_input_bits=-1, num_remat_bits=-1):
-        if num_input_bits > 0 or num_remat_bits > 0:
+        if num_input_bits > 0:
```
Comment (tgale96) [inline megablocks/layers/dmoe_test.py:149]: Can we just adjust the tolerance on allclose instead of bailing on the test?

Comment (dblalock) [reply megablocks/layers/dmoe_test.py:149]: So this is actually just skipping redundant test cases because whether we quantize the input or not doesn't affect any of the asserts here. Made the comment in the next line more explicit about this.

**Comment by dblalock** [comment]
Alright how about [this](https://api.wandb.ai/links/mosaic-ml/epefj2s1) results link?

File: megablocks/layers/dmoe_test.py (hunk 9 above)
Comment (tgale96) [reply megablocks/layers/dmoe_test.py:149]: Can we change the configs we pass to this test? This is kind of a nit, but I think that's the proper way to do it.

**Comment by tgale96** [comment]
Looks awesome! Once we resolve our convo on the unit tests we can go ahead and merge.

File: megablocks/layers/dmoe_test.py (hunk 9 above)
Comment (dblalock) [reply megablocks/layers/dmoe_test.py]: Done. The `_FORWARD_TESTS` are now split into more granular subsets.
